src/app/server/Server.cpp

```diff
@@ -70,6 +72,89 @@ constexpr bool useTestPairing()
 #endif
 }
 
+CHIP_ERROR PersistAdminPairingToKVS(AdminPairingInfo * admin, AdminId nextAvailableId)
+{
+    ReturnErrorCodeIf(admin == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    ChipLogProgress(AppServer, "Persisting admin ID %d, next available %d", admin->GetAdminId(), nextAvailableId);
+
+    ReturnErrorOnFailure(admin->StoreUsingKVSMgr(PersistedStorage::KeyValueStoreMgr()));
+    ReturnErrorOnFailure(PersistedStorage::KeyValueStoreMgr().Put(kAdminTableCountKey, &nextAvailableId, sizeof(nextAvailableId)));
+
+    ChipLogProgress(AppServer, "Persisting admin ID successfully");
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR RestoreAllAdminPairingsFromKVS(AdminPairingTable & adminPairings, AdminId & nextAvailableId)
+{
+    // It's not an error if the key doesn't exist. Just return right away.
+    VerifyOrReturnError(PersistedStorage::KeyValueStoreMgr().Get(kAdminTableCountKey, &nextAvailableId) == CHIP_NO_ERROR,
+                        CHIP_NO_ERROR);
+    ChipLogProgress(AppServer, "Next available admin ID is %d", nextAvailableId);
+
+    for (AdminId id = 0; id < nextAvailableId; id++)
+    {
+        AdminPairingInfo * admin = adminPairings.AssignAdminId(id);
+        // Recreate the binding if one exists in persistent storage. Else skip to the next ID
+        if (admin->FetchUsingKVSMgr(PersistedStorage::KeyValueStoreMgr()) != CHIP_NO_ERROR)
+        {
+            adminPairings.ReleaseAdminId(id);
+        }
+        else
+        {
+            ChipLogProgress(AppServer, "Found admin pairing for %d, node ID %llu", admin->GetAdminId(), admin->GetNodeId());
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void DeleteAdminPairingsInKVS(AdminId nextAvailableId)
+{
+    PersistedStorage::KeyValueStoreMgr().Delete(kAdminTableCountKey);
+    for (AdminId id = 0; id < nextAvailableId; id++)
+    {
+        AdminPairingInfo::DeleteUsingKVSMgr(PersistedStorage::KeyValueStoreMgr(), id);
+    }
+}
+
+CHIP_ERROR RestoreAllSessionsFromKVS(SecureSessionMgr & sessionMgr, RendezvousServer & server)
+{
+    uint16_t nextSessionKeyId = 0;
+    // It's not an error if the key doesn't exist. Just return right away.
+    VerifyOrReturnError(PersistedStorage::KeyValueStoreMgr().Get(kStorablePeerConnectionCountKey, &nextSessionKeyId) ==
+                            CHIP_NO_ERROR,
+                        CHIP_NO_ERROR);
+    ChipLogProgress(AppServer, "Found %d stored connections", nextSessionKeyId);
+
+    for (uint16_t keyId = 0; keyId < nextSessionKeyId; keyId++)
+    {
+        StorablePeerConnection connection;
+        if (CHIP_NO_ERROR == connection.FetchUsingKVSMgr(PersistedStorage::KeyValueStoreMgr(), keyId))
+        {
+            // The following is static to save on stack space usage
+            static PASESession session;
```

Comment:
I understand why this is necessary for memory-constrained platforms where we have to designate maximum stack sizes at compile time.  However, it makes this function non-reentrant, which is concerning (particularly in shared platform code).

I really wish we could target these sorts of optimizations only to those platforms where they are needed.  I also think there should be some type of protection here, or a loud warning in the comments that this function is not reentrant, or both.

For this particular case, the call from init IS safe.  A very simple mitigation would be to declare RestoreAllSessionsFromKVS static so nothing outside of Server.cpp can call it.

More generally, do we have a convention for calling these things out in shared platform code, at least as comments?  Have we decided which code paths and objects must be thread safe and which do not need to be?  If I'm running the server from a Linux node, I would strongly prefer my implementation not be encumbered by these sorts of optimizations.

Reply:
I agree. I am not too happy about this being static. It was only added when the device started crashing due to stack overflow. I can definitely make `RestoreAllSessionsFromKVS` static. It's not supposed to be used outside this file.

Reply:
This wastes this object forever just for startup.

How about allocating it using PlatformNew so that the memory can be reclaimed?